### PR TITLE
Fix Pager Node Issues

### LINKED ISSUE #2974: Pager Flow Layout is Not Aware of Programmatic Scrolls
`ASPagerFlowLayout` has a `currentIndexPath` property that it tries to keep up-to-date using the `targetContentOffsetForProposedContentOffset:` methods from `UICollectionViewLayout`.

The pager layout uses this index path to reset the `contentOffset` during animated bounds changes (rotation) so we stay looking at the same page.

Unfortunately, this method is not called for programmatic scrolls – only for updates, user scrolls, and animated bounds changes. 

In Pinterest, we have a screen where the header recedes when you scroll. If you programmatically change to a different page, and then trigger the header to recede, the pager flow layout detects an animated bounds change and scrolls over to its "currentIndexPath" which is now incorrect because we just programmatically scrolled away.

Two steps:
- Add a `scrollViewDidScroll:` call to the pager flow layout, and drive all `_currentIndexPath` updates from this call. This is the only hook we _always_ get from UIScrollView on scroll.
- Ignore animated bounds changes that only affect height, since all pagers scroll horizontally.

### PR DESCRIPTION
**Pull request by Adlai-Holler**
Resolves #2974 and other major pager node rotation/content inset issues.

- Deprecates `ASCollectionView.zeroContentInsets`.
  - This flag was added as a hack during development of `ASPagerNode` to "undo" the effects of `UIViewController.automaticallyAdjustsScrollViewInsets`.
  - The flag adds `contentInsets = .zero` into `layoutSubviews`, after `[super layoutSubviews]`
  - This means our `contentInsets` is thrashing around constantly.
  - The view controller _also_ sets `contentOffset` during its insets adjustment, which we didn't account for.
- Adds a flag `ASPagerNode.allowsAutomaticInsetsAdjustment`.
  - Defaults to `NO` since you virtually NEVER want UIVC to adjust pager node insets.
  - If `NO`, finds the owning view controller of the pager and sets `automaticallyAdjustsScrollViewInsets=NO`, and logs a message.
  - In the future, we may remove the automatic setting of this property and just log the message, so users retain control of their view controllers' flags.
  - If they for some reason _want_ automatic insets adjustment from their VC, they can set `allowsAutomaticInsetsAdjustment=YES` on their pager node.
- Adds a public category method `-[UIResponder asdk_associatedViewController]` that safely travels up the responder chain and finds the nearest view controller. This is used above, and it may very well be handy for users especially when debugging. I also put the view controller in the node's `-debugDescription` for fun. I could have left it out to keep this diff smaller.
- Adds a private class `ASResponderChainEnumerator` that is used in implementing above, and may come in handy in other places later.
- Replaces `ASPagerFlowLayout.currentIndexPath` with `ASPagerFlowLayout.currentCellNode` so that we can stay pointing at the same node across inserts/deletes.
- Makes `ASPagerFlowLayout` use `-shouldInvalidateLayoutForBoundsChange:` – which is like the `UICollectionViewLayout` version of `scrollViewDidScroll:` – to update its `currentIndexPath`, rather than using `targetContentOffsetForProposedContentOffset:withScrollingVelocity:`, which is not called during programmatic scrolls.
- When a collection view's bounds change, removes the `performBatchUpdates:` that we placed around `_relayoutAllNodes`. The stated reason for the batch updates was to get the relayout into the data controller's transaction pipeline, but it already _goes_ in that pipeline internally, and we don't need to submit an empty update to trigger a layout invalidation, since we're already manually invalidating the layout on the next line. This avoids triggering `UICollectionViewLayout.targetContentOffsetForProposedContentOffset:` needlessly, which happens on every collection view update, even empty ones.
- Fixes indentation throughout `ASPagerFlowLayout.m`.

In short:
Before:

After:

Note the flashing may look ugly here, but it's intentionally added by flow layout because in a real app, the layout transition is usually going to be bad.

### CODE REVIEW

**Comment by Adlai-Holler** [comment]
I know this diff is huge, and I attempted to decompose it into smaller updates but to get the correct pager behavior, all of these changes were necessary (excluding e.g. adding the view controller into node debugDescription which was just for fun and seems very low-risk).

File: AsyncDisplayKit/ASCollectionView.mm [hunk 1]
```diff
@@ -1911,19 +1924,11 @@ - (void)layer:(CALayer *)layer didChangeBoundsWithOldValue:(CGRect)oldBounds new
...
     if (changedInNonScrollingDirection) {
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-      // This actually doesn't perform an animation, but prevents the transaction block from being processed in the
-      // data controller's prevent animation block that would interrupt an interrupted relayout happening in an animation block
-      // ie. ASCollectionView bounds change on rotation or multi-tasking split view resize.
-      [self performBatchAnimated:YES updates:^{
-        [_dataController relayoutAllNodes];
-      } completion:nil];
+      [_dataController relayoutAllNodes];
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASCollectionView.mm]: So much clearer.

File: AsyncDisplayKit/ASPagerFlowLayout.m [hunk 2]
```diff
@@ -75,19 +56,49 @@ - (CGPoint)_targetContentOffsetForItemAtIndexPath:(NSIndexPath *)indexPath propo
...
+  // isn't sufficient here! It's broken!
+  NSArray *animKeys = self.collectionView.layer.animationKeys;
+  for (NSString *key in animKeys) {
+    if ([key hasPrefix:@"bounds"]) {
+      return;
+    }
+  }
+  
+  CGRect bounds = self.collectionView.bounds;
+  CGRect rect = CGRectMake(CGRectGetMidX(bounds), CGRectGetMidY(bounds), 1, 1);
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerFlowLayout.m]: Probably doesn't matter, but should this possibly be `CGRectMake(CGRectGetMidX(bounds) - 0.5, CGRectGetMidY(bounds) - 0.5, 1, 1);`

Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerFlowLayout.m]: I assume it doesn't because the pager should have cells that fill the view?

File: AsyncDisplayKit/ASPagerNode.h [hunk 3]
```diff
@@ -120,6 +120,24 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ * @discussion ASPagerNode should usually not have its content insets automatically adjusted
+ * because it scrolls horizontally, and flow layout will log errors because the pages
+ * do not fit between the top & bottom insets of the collection view.
+ *
+ * The default value is NO, which means that ASPagerNode expects that its view controller will
+ * have automaticallyAdjustsScrollViewInsets=NO.
+ *
+ * If this property is NO, but your view controller has automaticallyAdjustsScrollViewInsets=YES,
+ * the pager node will set the property to NO and log a warning message. In the future,
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerNode.h]: This isn't 100% clear to me. It will set the property on the view controller to NO? Will it **leave** the pageNode property NO? Or, will it actually set the pagerNode property to YES?

File: AsyncDisplayKit/ASPagerNode.m [hunk 4]
```diff
@@ -211,4 +208,21 @@ - (void)proxyTargetHasDeallocated:(ASDelegateProxy *)proxy
   [self setDelegate:nil];
 }
 
+- (void)didEnterVisibleState
+{
+	[super didEnterVisibleState];
+
+	// Check that our view controller does not automatically set our content insets
+	// It would be better to have a -didEnterHierarchy hook to put this in, but
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerNode.m]: Will `willEnterHierarchy` not suffice?

File: AsyncDisplayKit/UIResponder+AsyncDisplayKit.m [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#import <AsyncDisplayKit/ASResponderChainEnumerator.h>
+
+@implementation UIResponder (AsyncDisplayKit)
+
+- (__kindof UIViewController *)asdk_associatedViewController
+{
+  ASDisplayNodeAssertMainThread();
+  
+  for (UIResponder *responder in [self asdk_responderChainEnumerator]) {
+    UIViewController *vc = ASDynamicCast(responder, UIViewController);
```
Comment (garrettmoon) [inline AsyncDisplayKit/UIResponder+AsyncDisplayKit.m]: Is this documented behavior? A responder will have it's viewController in the responder chain prior to any other possible view controllers at all times? If not, maybe a comment indicating it's been tested to be true in practice? If so a link would be awesome :)

File: examples/PagerNode/Sample/PageNode.m [hunk 6]
```diff
@@ -21,9 +21,9 @@
 
 @implementation PageNode
 
-- (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
+- (CGSize)calculateSizeThatFits:(CGSize)constrainedSize
```
Comment (garrettmoon) [inline examples/PagerNode/Sample/PageNode.m]: `@note Subclasses that override are committed to manual layout. Therefore, -layout: must be overriden to layout all subnodes or subviews.`

Not needed in this case?

File: AsyncDisplayKit/ASPagerFlowLayout.m (hunk 2 above)
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASPagerFlowLayout.m]: Right – "government work" level precision is enough here 🏗

File: AsyncDisplayKit/ASPagerNode.h (hunk 3 above)
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASPagerNode.h]: It will leave the pager property (as NO) but it'll automatically set NO on the view controller (for backwards compatibility) and log a message that it did so. In the future, we want users to be responsible for configuring their own view controllers, so we'll just log the message.

File: AsyncDisplayKit/ASPagerNode.m (hunk 4 above)
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASPagerNode.m]: No unfortunately, because we're not yet in the hierarchy =(

File: AsyncDisplayKit/UIResponder+AsyncDisplayKit.m (hunk 5 above)
Comment (Adlai-Holler) [inline AsyncDisplayKit/UIResponder+AsyncDisplayKit.m]: Hmm I'll leave the link out because the best I could do is a general link to an overview of the responder chain. But responder chain always goes `aView -> aView.superview -> aView.superview.viewController -> aView.superview.viewController.view.superview` etc, which is documented.

File: examples/PagerNode/Sample/PageNode.m (hunk 6 above)
Comment (Adlai-Holler) [inline examples/PagerNode/Sample/PageNode.m]: Right because we have no subnodes.

File: AsyncDisplayKit/ASPagerNode.m (hunk 4 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerNode.m]: Interesting, the isInHierarchy flag is set before willEnterHierarchy is called…

File: AsyncDisplayKit/UIResponder+AsyncDisplayKit.m (hunk 5 above)
Comment (garrettmoon) [inline AsyncDisplayKit/UIResponder+AsyncDisplayKit.m]: 👍

File: AsyncDisplayKit/ASPagerNode.h (hunk 3 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerNode.h]: Maybe just a slight adjustment to your comment then: `If this property is NO, but your view controller has automaticallyAdjustsScrollViewInsets=YES, the pager node will set the property **on the view controller** to NO and log a warning message.`

Comment (garrettmoon) [inline AsyncDisplayKit/ASPagerNode.h]: But without the stars :)
